[Cleanup] Fix issues relating to ambiguous virtual function overloads

peter-b opened the pull request:
When a base class declares a method `virtual X::f(int)`, and a derived class declares a method `virtual Y::f(void)`, the derived class's incompatibly overloaded `f()` "masks" the base class's `f()`.  It makes it impossibly to call `Y y; Y.f(0)`, for example; an explicit cast to `X` would be required.  In these situations, recent versions of clang (such as the Emscripten compiler) provides a "overloaded-virtual" warning.

This pull request fixes all "overloaded-virtual" warnings in the Emscripten build, and turns "overloaded-virtual" warnings into errors.

peter-b committed:
MCObject::load() shouldn't be called for MCStack instances

When loading a stack, the type byte needs to be passed to
`MCStack::load()` in order to allow the correct stack loading mode to
be selected.  All other objects can be fully loaded without the type
byte.

Previously, only a 3-argument variant of `load()` was defined in
`MCStack`, which meant that it was possible to (incorrectly) call the
base class (2-argument) implementation of `MCObject::load()` by
upcasting an `MCStack` pointer.  This was flagged up by the emscripten
C++ compiler with a "overloaded-virtual" warning.

This patch explicitly overrides `MCObject::load()` in `MCStack` with
an implementation that indicates a programming error.

peter-b committed:
[Cleanup] Rename stack "Shadow" accessors to "WindowShadow".

`MCObject::SetShadow(MCExecContext &, const MCInterfaceShadow &)` is a
virtual method.  Originally, it was masked by the overloaded virtual
method `MCStack::SetShadow(MCExecContext &, bool)`, which does
something completely different.  In order to avoid ambiguity, either
`MCStack` needs a `using` declaration for `MCObject::SetShadow`, or
`MCStack::SetShadow` needs to be renamed.

This patch renames `MCStack::SetShadow` to `MCStack::SetWindowShadow`
(and likewise for `GetShadow`).

peter-b committed:
[emscripten] Add -Werror=overloaded-virtual.

Turn "overloaded-virtual" warnings into errors when compiling for
emscripten.

peter-b commented:
@livecode-vulcan review not ok 4609755

Failed tests locally.

livecode-vulcan commented:
💔 review by @peter-b not ok 4609755d44efff5cc154e91572b7a3af13471147

peter-b committed:
Disambiguate MCUIDC::dodragdrop() overloads on mobile.

`MCUIDC` provides a `dodragdrop()` method (with a do-nothing default
implementation), and the mobile `MCScreenDC` subclass overloads it
with a different do-nothing implementation).  Add an explicit `using`
to the mobile `MCScreenDC` in order to make it unambiguous that both
overloads of `dodragdrop()` are present (and do nothing).

peter-b commented:
Pull request updated.

runrevmark commented on engine/src/mbldc.h:
```diff
@@ -195,7 +195,11 @@ class MCScreenDC: public MCUIDC
+	 * which uses a pasteboard. By default, both do nothing
+	 * (i.e. always return DRAG_ACTION_NONE). */
+	using MCUIDC::dodragdrop;
+	virtual MCDragAction dodragdrop(Window w, MCPasteboard *p_pasteboard, MCDragActionSet p_allowed_actions, MCImage *p_image, const MCPoint* p_image_offset);
```
This shouldn't be necessary - more likely the mobile signature just hasn't been updated. DragDrop doesn't exist on mobile.

peter-b committed:
Remove mobile-specific MCUIDC::dodragdrop() overload.

As pointed out by @runrevmark, commit f419a9d53720 wasn't the correct
fix for the "overloaded-virtual" warning.  Since drag & drop isn't
actually supported on mobile, the mobile `MCScreenDC::dodragdrop()`
method should actually be removed entirely.

peter-b committed:
Merge branch 'develop' into misc/cleanup-stack-shadow

Conflict in `config/emscripten-settings.gypi` because of conflicting
additions of `-Werror` flags.

peter-b commented:
@runrevmark Pull request updated to remove that `dodragdrop()` method entirely.

runrevmark commented:
@livecode-vulcan review ok 00e38b1

livecode-vulcan commented:
💙 review by @runrevmark ok 00e38b1f8d6f91e530da187f8308d98744f5bb2f

livecode-vulcan commented:
😞 test failure 00e38b1f8d6f91e530da187f8308d98744f5bb2f
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1046)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1055)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1043)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1048)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1058)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1064)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1063)

peter-b commented:
@livecode-vulcan retry 00e38b1

livecode-vulcan commented:
😎 test success 00e38b1f8d6f91e530da187f8308d98744f5bb2f
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1052)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1061)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1049)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1054)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1064)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1069)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1069)
